src/main/java/io/reactivex/internal/operators/flowable/FlowableThrottleAndSample.java

```diff
@@ -0,0 +1,221 @@
+/**
+ * Copyright (c) 2016-present, RxJava Contributors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in
+ * compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is
+ * distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See
+ * the License for the specific language governing permissions and limitations under the License.
+ */
+package io.reactivex.internal.operators.flowable;
+
+import io.reactivex.Flowable;
+import io.reactivex.FlowableSubscriber;
+import io.reactivex.Scheduler;
+import io.reactivex.annotations.NonNull;
+import io.reactivex.disposables.Disposables;
+import io.reactivex.exceptions.MissingBackpressureException;
+import io.reactivex.internal.disposables.DisposableHelper;
+import io.reactivex.internal.disposables.SequentialDisposable;
+import io.reactivex.internal.subscriptions.SubscriptionHelper;
+import io.reactivex.internal.util.BackpressureHelper;
+import io.reactivex.plugins.RxJavaPlugins;
+import io.reactivex.subscribers.SerializedSubscriber;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.*;
+
+public final class FlowableThrottleAndSample<T> extends AbstractFlowableWithUpstream<T, T> {
+
+    final long windowDuration;
+    final TimeUnit unit;
+    final Scheduler scheduler;
+
+    final boolean emitLast;
+
+    public FlowableThrottleAndSample(@NonNull final Flowable<T> source, final long windowDuration,
+                                     final TimeUnit unit, final Scheduler scheduler, final boolean emitLast) {
+        super(source);
+        this.windowDuration = windowDuration;
+        this.unit = unit;
+        this.scheduler = scheduler;
+        this.emitLast = emitLast;
+    }
+
+    @Override
+    protected void subscribeActual(final Subscriber<? super T> s) {
+        final SerializedSubscriber<T> serial = new SerializedSubscriber<T>(s);
+        if (emitLast) {
+            source.subscribe(new ThrottleAndSampleEmitLast<T>(serial, windowDuration, unit, scheduler.createWorker()));
+        } else {
+            source.subscribe(new ThrottleAndSampleNoLast<T>(serial, windowDuration, unit, scheduler.createWorker()));
+        }
+    }
+
+    abstract static class ThrottleAndSampleSubscriber<T> extends AtomicReference<T> implements FlowableSubscriber<T>, Subscription, Runnable {
```

Comment:
Accessing `timerInProgress` happens more often so `extends AtomicInteger` is better. Have the `AtomicReference` as instance field instead.